{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('..')\n",
    "from MPRA_predict.utils import *\n",
    "\n",
    "os.chdir('../data/TewheyMPRA')\n",
    "\n",
    "from pyfaidx import Fasta\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cell_type = 'GM12878'\n",
    "# cell_type = 'A549'\n",
    "cell_type = 'HCT116'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('OL41B', 242177), ('ctrl_neg', 506), ('ctrl_pos', 91)]\n",
      "3.0855209378503883 0.17389626743795658 0.3594951115071403\n",
      "[('OL42', 251221), ('ctrl_neg', 506), ('ctrl_pos', 91)]\n",
      "3.016368974717195 0.15183169335106786 0.5271926355812913\n",
      "406808 406808\n"
     ]
    }
   ],
   "source": [
    "df_list = []\n",
    "for file_name in os.listdir(cell_type):\n",
    "    if file_name.endswith('.tsv'):\n",
    "        df = pd.read_csv(f'{cell_type}/{file_name}', sep='\\t', low_memory=False)\n",
    "\n",
    "        c = Counter(df['ctrl_exp'])\n",
    "        print(c.most_common())\n",
    "        df['experiment'] = c.most_common()[0][0]\n",
    "        ctrl_pos_mean = df[df['ctrl_exp'] == 'ctrl_pos']['log2FoldChange'].mean()\n",
    "        ctrl_neg_mean = df[df['ctrl_exp'] == 'ctrl_neg']['log2FoldChange'].mean()\n",
    "        total_mean = df['log2FoldChange'].mean()\n",
    "        print(ctrl_pos_mean, ctrl_neg_mean, total_mean)\n",
    "        df['log2FoldChangeOverControl'] = (df['log2FoldChange'] - ctrl_neg_mean) / (ctrl_pos_mean - ctrl_neg_mean)\n",
    "        df_list.append(df)\n",
    "df = pd.concat(df_list).reset_index(drop=True)\n",
    "df\n",
    "\n",
    "# 不能简单合并，要按照ctrl做归一化再合并\n",
    "# 这里把ctrl_pos归一化到1，把ctrl_neg归一化到0，计算log2FoldChangeOverControl\n",
    "\n",
    "\n",
    "\n",
    "df['chr'] = df['chr'].apply(lambda x: 'chr' + x)\n",
    "df.rename(columns={\n",
    "    'ref_allele': 'ref', \n",
    "    'alt_allele': 'alt', \n",
    "    }, inplace=True)\n",
    "\n",
    "df = df.sort_values(by=['chr', 'pos']).reset_index(drop=True)\n",
    "\n",
    "# 只保留ID只有5个:的\n",
    "df = df[df['ID'].str.count(':') == 5].reset_index(drop=True)\n",
    "# 只保留window是center的\n",
    "df = df[df['window'] == 'center'].reset_index(drop=True)\n",
    "# 只保留allele长度为1的或者allele为nan的\n",
    "df = df[((df['ref'].str.len() == 1) & (df['alt'].str.len() == 1)) | (df['ref'].isna())].reset_index(drop=True)\n",
    "df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# 合并相同ID的序列，活性使用平均值\n",
    "# df = df.groupby(['ID'], as_index=False).agg(lambda x: x.mean() if x.dtype == float else x.iloc[0]).reset_index(drop=True)\n",
    "df = df.groupby('ID', as_index=False).agg(\n",
    "    {col: 'mean' if df[col].dtype == 'float' else 'first' for col in df.columns}\n",
    ")\n",
    "# 删除重复'chr', 'pos', 'allele'的序列，注意这些序列的ID不同，所以是同一个位点，做了两次突变，生成了4条序列的情况，直接删除后两条\n",
    "df = df.drop_duplicates(subset=['chr', 'pos', 'allele'])\n",
    "df = df.sort_values(by=['chr', 'pos', 'allele']).reset_index(drop=True)\n",
    "df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "genome = Fasta('/home/hxcai/genome/hg19.fa')\n",
    "\n",
    "def process_row(row):\n",
    "    chr, pos, ref, alt, allele = row['chr'], row['pos'], row['ref'], row['alt'], row['allele']\n",
    "\n",
    "    genome_ref_base = genome[chr][pos-1].seq.upper()\n",
    "    correct_flag = genome_ref_base == ref  # 位置正确的标志\n",
    "\n",
    "    seq = genome[chr][pos-100:pos+100].seq.upper()\n",
    "    if allele != ref:\n",
    "        seq = seq[:99] + alt + seq[100:]\n",
    "\n",
    "    return seq, correct_flag\n",
    "\n",
    "df[['seq', 'flags']] = df.apply(process_row, axis=1, result_type='expand')\n",
    "print(len(df), df['flags'].sum())\n",
    "\n",
    "\n",
    "\n",
    "df = df[['ID', 'chr', 'pos', 'allele', 'seq', 'log2FoldChange', 'log2FoldChangeOverControl']]\n",
    "df.to_csv(f'TewheyMPRA_{cell_type}.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "torch",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
