{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('..')\n",
    "from MPRA_predict.utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>log2FoldChange_expr</th>\n",
       "      <th>lfcSE_expr</th>\n",
       "      <th>stat_expr</th>\n",
       "      <th>pvalue_expr</th>\n",
       "      <th>padj_expr</th>\n",
       "      <th>log2FoldChange_allele</th>\n",
       "      <th>lfcSE_allele</th>\n",
       "      <th>stat_allele</th>\n",
       "      <th>pvalue_allele</th>\n",
       "      <th>padj_allele</th>\n",
       "      <th>bin</th>\n",
       "      <th>score</th>\n",
       "      <th>negative</th>\n",
       "      <th>regulatory_hit</th>\n",
       "      <th>allelic_hit</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>sign_concord</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tss_dist</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seq_10005_ENSG00000015133</td>\n",
       "      <td>0.397</td>\n",
       "      <td>0.080</td>\n",
       "      <td>4.979</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.001</td>\n",
       "      <td>0.159</td>\n",
       "      <td>-0.007</td>\n",
       "      <td>0.995</td>\n",
       "      <td>0.997</td>\n",
       "      <td>bin_3</td>\n",
       "      <td>-0.023</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>chr14</td>\n",
       "      <td>91417778</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>41</td>\n",
       "      <td>ENSG00000015133</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>seq_10010_ENSG00000170113</td>\n",
       "      <td>1.221</td>\n",
       "      <td>0.275</td>\n",
       "      <td>4.442</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.669</td>\n",
       "      <td>0.548</td>\n",
       "      <td>-1.220</td>\n",
       "      <td>0.222</td>\n",
       "      <td>0.476</td>\n",
       "      <td>bin_3</td>\n",
       "      <td>-0.107</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>chr15</td>\n",
       "      <td>22786558</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>-85</td>\n",
       "      <td>ENSG00000170113</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seq_10011_ENSG00000140092</td>\n",
       "      <td>0.489</td>\n",
       "      <td>0.041</td>\n",
       "      <td>11.912</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.120</td>\n",
       "      <td>0.082</td>\n",
       "      <td>-1.463</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.365</td>\n",
       "      <td>bin_3</td>\n",
       "      <td>0.018</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr14</td>\n",
       "      <td>91947694</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000140092</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>seq_10016_ENSG00000184058</td>\n",
       "      <td>2.883</td>\n",
       "      <td>0.453</td>\n",
       "      <td>6.366</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-1.576</td>\n",
       "      <td>0.903</td>\n",
       "      <td>-1.745</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.257</td>\n",
       "      <td>bin_3</td>\n",
       "      <td>0.108</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr22</td>\n",
       "      <td>19760739</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-38</td>\n",
       "      <td>ENSG00000184058</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>seq_10019_ENSG00000185883</td>\n",
       "      <td>2.773</td>\n",
       "      <td>0.028</td>\n",
       "      <td>99.885</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.138</td>\n",
       "      <td>0.055</td>\n",
       "      <td>-2.488</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.066</td>\n",
       "      <td>bin_3</td>\n",
       "      <td>-0.003</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2513844</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-107</td>\n",
       "      <td>ENSG00000185883</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13793</th>\n",
       "      <td>seq_9981_ENSG00000128340</td>\n",
       "      <td>-0.225</td>\n",
       "      <td>0.037</td>\n",
       "      <td>-6.132</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.103</td>\n",
       "      <td>0.073</td>\n",
       "      <td>-1.406</td>\n",
       "      <td>0.160</td>\n",
       "      <td>0.433</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>-0.086</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>chr22</td>\n",
       "      <td>37244475</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-207</td>\n",
       "      <td>ENSG00000128340</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13794</th>\n",
       "      <td>seq_9989_ENSG00000103051</td>\n",
       "      <td>-0.430</td>\n",
       "      <td>0.043</td>\n",
       "      <td>-10.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.090</td>\n",
       "      <td>0.086</td>\n",
       "      <td>1.049</td>\n",
       "      <td>0.294</td>\n",
       "      <td>0.593</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>-0.011</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70523117</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>436</td>\n",
       "      <td>ENSG00000103051</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13795</th>\n",
       "      <td>seq_998_ENSG00000189057</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.319</td>\n",
       "      <td>0.750</td>\n",
       "      <td>0.764</td>\n",
       "      <td>0.065</td>\n",
       "      <td>0.051</td>\n",
       "      <td>1.273</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.496</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>-0.009</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr11</td>\n",
       "      <td>59107666</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>430</td>\n",
       "      <td>ENSG00000189057</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13796</th>\n",
       "      <td>seq_9998_ENSG00000155961</td>\n",
       "      <td>0.304</td>\n",
       "      <td>0.051</td>\n",
       "      <td>5.926</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.090</td>\n",
       "      <td>0.103</td>\n",
       "      <td>-0.880</td>\n",
       "      <td>0.379</td>\n",
       "      <td>0.679</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155264059</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>431</td>\n",
       "      <td>ENSG00000155961</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13797</th>\n",
       "      <td>seq_999_ENSG00000166801</td>\n",
       "      <td>0.099</td>\n",
       "      <td>0.036</td>\n",
       "      <td>2.766</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.007</td>\n",
       "      <td>-0.030</td>\n",
       "      <td>0.071</td>\n",
       "      <td>-0.418</td>\n",
       "      <td>0.676</td>\n",
       "      <td>0.868</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>0.002</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr11</td>\n",
       "      <td>59142561</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-297</td>\n",
       "      <td>ENSG00000166801</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13798 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              ID  log2FoldChange_expr  lfcSE_expr  stat_expr  pvalue_expr  padj_expr  log2FoldChange_allele  lfcSE_allele  stat_allele  pvalue_allele  padj_allele    bin  score  negative  regulatory_hit  allelic_hit  high_confidence  sign_concord  chrom        pos ref alt  tss_dist          gene_id     category\n",
       "0      seq_10005_ENSG00000015133                0.397       0.080      4.979        0.000      0.000                 -0.001         0.159       -0.007          0.995        0.997  bin_3 -0.023     False            True        False            False          True  chr14   91417778   G   A        41  ENSG00000015133  monoallelic\n",
       "1      seq_10010_ENSG00000170113                1.221       0.275      4.442        0.000      0.000                 -0.669         0.548       -1.220          0.222        0.476  bin_3 -0.107     False            True        False            False          True  chr15   22786558   T   G       -85  ENSG00000170113  monoallelic\n",
       "2      seq_10011_ENSG00000140092                0.489       0.041     11.912        0.000      0.000                 -0.120         0.082       -1.463          0.143        0.365  bin_3  0.018     False            True        False            False         False  chr14   91947694   C   A        -1  ENSG00000140092  monoallelic\n",
       "3      seq_10016_ENSG00000184058                2.883       0.453      6.366        0.000      0.000                 -1.576         0.903       -1.745          0.081        0.257  bin_3  0.108     False            True        False            False         False  chr22   19760739   G   A       -38  ENSG00000184058  monoallelic\n",
       "4      seq_10019_ENSG00000185883                2.773       0.028     99.885        0.000      0.000                 -0.138         0.055       -2.488          0.013        0.066  bin_3 -0.003     False            True        False            False          True  chr16    2513844   C   T      -107  ENSG00000185883  monoallelic\n",
       "...                          ...                  ...         ...        ...          ...        ...                    ...           ...          ...            ...          ...    ...    ...       ...             ...          ...              ...           ...    ...        ...  ..  ..       ...              ...          ...\n",
       "13793   seq_9981_ENSG00000128340               -0.225       0.037     -6.132        0.000      0.000                 -0.103         0.073       -1.406          0.160        0.433  bin_1 -0.086     False            True        False            False          True  chr22   37244475   G   A      -207  ENSG00000128340  monoallelic\n",
       "13794   seq_9989_ENSG00000103051               -0.430       0.043    -10.003        0.000      0.000                  0.090         0.086        1.049          0.294        0.593  bin_1 -0.011     False            True        False            False         False  chr16   70523117   G   A       436  ENSG00000103051  monoallelic\n",
       "13795    seq_998_ENSG00000189057                0.008       0.025      0.319        0.750      0.764                  0.065         0.051        1.273          0.203        0.496  bin_1 -0.009      True           False        False            False         False  chr11   59107666   G   C       430  ENSG00000189057  monoallelic\n",
       "13796   seq_9998_ENSG00000155961                0.304       0.051      5.926        0.000      0.000                 -0.090         0.103       -0.880          0.379        0.679  bin_1  0.000     False            True        False            False         False   chrX  155264059   G   A       431  ENSG00000155961  monoallelic\n",
       "13797    seq_999_ENSG00000166801                0.099       0.036      2.766        0.006      0.007                 -0.030         0.071       -0.418          0.676        0.868  bin_1  0.002     False            True        False            False         False  chr11   59142561   C   T      -297  ENSG00000166801  monoallelic\n",
       "\n",
       "[13798 rows x 25 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../data/promoterAI_MPRA/tableS6.tsv', sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>log2FoldChange_expr</th>\n",
       "      <th>lfcSE_expr</th>\n",
       "      <th>stat_expr</th>\n",
       "      <th>pvalue_expr</th>\n",
       "      <th>padj_expr</th>\n",
       "      <th>log2FoldChange_allele</th>\n",
       "      <th>lfcSE_allele</th>\n",
       "      <th>stat_allele</th>\n",
       "      <th>pvalue_allele</th>\n",
       "      <th>padj_allele</th>\n",
       "      <th>bin</th>\n",
       "      <th>score</th>\n",
       "      <th>negative</th>\n",
       "      <th>regulatory_hit</th>\n",
       "      <th>allelic_hit</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>sign_concord</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tss_dist</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10293</th>\n",
       "      <td>seq_12656_ENSG00000178591</td>\n",
       "      <td>0.554</td>\n",
       "      <td>0.027</td>\n",
       "      <td>20.571</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.054</td>\n",
       "      <td>0.440</td>\n",
       "      <td>0.660</td>\n",
       "      <td>0.859</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>-0.005</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr20</td>\n",
       "      <td>87220</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-451</td>\n",
       "      <td>ENSG00000178591</td>\n",
       "      <td>DNM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10474</th>\n",
       "      <td>seq_13201_ENSG00000250312</td>\n",
       "      <td>0.204</td>\n",
       "      <td>0.033</td>\n",
       "      <td>6.192</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.066</td>\n",
       "      <td>0.274</td>\n",
       "      <td>0.784</td>\n",
       "      <td>0.912</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>0.013</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>chr4</td>\n",
       "      <td>124152</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-348</td>\n",
       "      <td>ENSG00000250312</td>\n",
       "      <td>DNM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1539</th>\n",
       "      <td>seq_2399_ENSG00000103148</td>\n",
       "      <td>-0.174</td>\n",
       "      <td>0.023</td>\n",
       "      <td>-7.523</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.073</td>\n",
       "      <td>0.046</td>\n",
       "      <td>1.577</td>\n",
       "      <td>0.115</td>\n",
       "      <td>0.320</td>\n",
       "      <td>bin_3</td>\n",
       "      <td>-0.041</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr16</td>\n",
       "      <td>138351</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>321</td>\n",
       "      <td>ENSG00000103148</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11542</th>\n",
       "      <td>seq_2400_ENSG00000103148</td>\n",
       "      <td>0.965</td>\n",
       "      <td>0.018</td>\n",
       "      <td>53.224</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.009</td>\n",
       "      <td>0.036</td>\n",
       "      <td>-0.257</td>\n",
       "      <td>0.797</td>\n",
       "      <td>0.919</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>0.049</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr16</td>\n",
       "      <td>138945</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-273</td>\n",
       "      <td>ENSG00000103148</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11543</th>\n",
       "      <td>seq_2401_ENSG00000103148</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.033</td>\n",
       "      <td>20.678</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.096</td>\n",
       "      <td>0.066</td>\n",
       "      <td>-1.446</td>\n",
       "      <td>0.148</td>\n",
       "      <td>0.414</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>-0.007</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>chr16</td>\n",
       "      <td>139075</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-403</td>\n",
       "      <td>ENSG00000103148</td>\n",
       "      <td>monoallelic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9462</th>\n",
       "      <td>seq_10527_ENSG00000198452</td>\n",
       "      <td>0.506</td>\n",
       "      <td>0.027</td>\n",
       "      <td>18.715</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.045</td>\n",
       "      <td>0.054</td>\n",
       "      <td>-0.834</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.701</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>-0.037</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>247616986</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>-156</td>\n",
       "      <td>ENSG00000198452</td>\n",
       "      <td>DNM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4177</th>\n",
       "      <td>seq_10529_ENSG00000238243</td>\n",
       "      <td>-0.263</td>\n",
       "      <td>0.046</td>\n",
       "      <td>-5.665</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.103</td>\n",
       "      <td>0.093</td>\n",
       "      <td>1.111</td>\n",
       "      <td>0.267</td>\n",
       "      <td>0.596</td>\n",
       "      <td>bin_2</td>\n",
       "      <td>-0.013</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>247896081</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>495</td>\n",
       "      <td>ENSG00000238243</td>\n",
       "      <td>DNM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9463</th>\n",
       "      <td>seq_10530_ENSG00000196071</td>\n",
       "      <td>-0.366</td>\n",
       "      <td>0.117</td>\n",
       "      <td>-3.135</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.002</td>\n",
       "      <td>-0.072</td>\n",
       "      <td>0.234</td>\n",
       "      <td>-0.310</td>\n",
       "      <td>0.756</td>\n",
       "      <td>0.903</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>0.002</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248096814</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-461</td>\n",
       "      <td>ENSG00000196071</td>\n",
       "      <td>DNM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9464</th>\n",
       "      <td>seq_10531_ENSG00000196240</td>\n",
       "      <td>-0.009</td>\n",
       "      <td>0.041</td>\n",
       "      <td>-0.225</td>\n",
       "      <td>0.822</td>\n",
       "      <td>0.832</td>\n",
       "      <td>-0.078</td>\n",
       "      <td>0.081</td>\n",
       "      <td>-0.963</td>\n",
       "      <td>0.335</td>\n",
       "      <td>0.636</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>-0.001</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248446123</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>487</td>\n",
       "      <td>ENSG00000196240</td>\n",
       "      <td>DNM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9465</th>\n",
       "      <td>seq_10532_ENSG00000259823</td>\n",
       "      <td>-0.131</td>\n",
       "      <td>0.066</td>\n",
       "      <td>-1.991</td>\n",
       "      <td>0.047</td>\n",
       "      <td>0.053</td>\n",
       "      <td>-0.198</td>\n",
       "      <td>0.132</td>\n",
       "      <td>-1.502</td>\n",
       "      <td>0.133</td>\n",
       "      <td>0.392</td>\n",
       "      <td>bin_1</td>\n",
       "      <td>0.024</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248755552</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>206</td>\n",
       "      <td>ENSG00000259823</td>\n",
       "      <td>DNM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13798 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              ID  log2FoldChange_expr  lfcSE_expr  stat_expr  pvalue_expr  padj_expr  log2FoldChange_allele  lfcSE_allele  stat_allele  pvalue_allele  padj_allele    bin  score  negative  regulatory_hit  allelic_hit  high_confidence  sign_concord  chrom        pos ref alt  tss_dist          gene_id     category\n",
       "10293  seq_12656_ENSG00000178591                0.554       0.027     20.571        0.000      0.000                  0.024         0.054        0.440          0.660        0.859  bin_1 -0.005     False            True        False            False         False  chr20      87220   T   C      -451  ENSG00000178591          DNM\n",
       "10474  seq_13201_ENSG00000250312                0.204       0.033      6.192        0.000      0.000                  0.018         0.066        0.274          0.784        0.912  bin_1  0.013     False            True        False            False          True   chr4     124152   T   C      -348  ENSG00000250312          DNM\n",
       "1539    seq_2399_ENSG00000103148               -0.174       0.023     -7.523        0.000      0.000                  0.073         0.046        1.577          0.115        0.320  bin_3 -0.041     False            True        False            False         False  chr16     138351   C   G       321  ENSG00000103148  monoallelic\n",
       "11542   seq_2400_ENSG00000103148                0.965       0.018     53.224        0.000      0.000                 -0.009         0.036       -0.257          0.797        0.919  bin_1  0.049     False            True        False            False         False  chr16     138945   C   T      -273  ENSG00000103148  monoallelic\n",
       "11543   seq_2401_ENSG00000103148                0.688       0.033     20.678        0.000      0.000                 -0.096         0.066       -1.446          0.148        0.414  bin_1 -0.007     False            True        False            False          True  chr16     139075   A   G      -403  ENSG00000103148  monoallelic\n",
       "...                          ...                  ...         ...        ...          ...        ...                    ...           ...          ...            ...          ...    ...    ...       ...             ...          ...              ...           ...    ...        ...  ..  ..       ...              ...          ...\n",
       "9462   seq_10527_ENSG00000198452                0.506       0.027     18.715        0.000      0.000                 -0.045         0.054       -0.834          0.404        0.701  bin_1 -0.037     False            True        False            False          True   chr1  247616986   A   T      -156  ENSG00000198452          DNM\n",
       "4177   seq_10529_ENSG00000238243               -0.263       0.046     -5.665        0.000      0.000                  0.103         0.093        1.111          0.267        0.596  bin_2 -0.013     False            True        False            False         False   chr1  247896081   C   G       495  ENSG00000238243          DNM\n",
       "9463   seq_10530_ENSG00000196071               -0.366       0.117     -3.135        0.002      0.002                 -0.072         0.234       -0.310          0.756        0.903  bin_1  0.002     False            True        False            False         False   chr1  248096814   G   A      -461  ENSG00000196071          DNM\n",
       "9464   seq_10531_ENSG00000196240               -0.009       0.041     -0.225        0.822      0.832                 -0.078         0.081       -0.963          0.335        0.636  bin_1 -0.001      True           False        False            False          True   chr1  248446123   T   C       487  ENSG00000196240          DNM\n",
       "9465   seq_10532_ENSG00000259823               -0.131       0.066     -1.991        0.047      0.053                 -0.198         0.132       -1.502          0.133        0.392  bin_1  0.024     False           False        False            False         False   chr1  248755552   T   G       206  ENSG00000259823          DNM\n",
       "\n",
       "[13798 rows x 25 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_values(by=['pos'], inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = Fasta('../../../genome/hg38.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 42%|████▏     | 801/1896 [00:00<00:00, 8008.37it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1896/1896 [00:00<00:00, 3929.77it/s]\n",
      "100%|██████████| 1896/1896 [00:00<00:00, 6044.10it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>seq</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>allele</th>\n",
       "      <th>casual_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_63671_G_A_b38</td>\n",
       "      <td>GGGCACCTTCTTTTTCTTAATTGTATCATACATTTTTATTCTGGTC...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>63671</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ref</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_63671_G_A_b38</td>\n",
       "      <td>GGGCACCTTCTTTTTCTTAATTGTATCATACATTTTTATTCTGGTC...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>63671</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>alt</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_108826_G_C_b38</td>\n",
       "      <td>TGGATCCTTAATAAACAAGGATGTTTCTGCATCATTTCCCCACAAC...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108826</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>ref</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_108826_G_C_b38</td>\n",
       "      <td>TGGATCCTTAATAAACAAGGATGTTTCTGCATCATTTCCCCACAAC...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108826</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>alt</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_135203_G_A_b38</td>\n",
       "      <td>GGAGCTTTGGACTGGGGAGGCCGCAGTGAGGCGAGAGCTAGCTGGG...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>135203</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ref</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7579</th>\n",
       "      <td>chrX_155668742_C_A_b38</td>\n",
       "      <td>TTTGAGCCTGCTTTAAACCATGAAATTCAACACCATCCTCAGATTC...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155668742</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>alt</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7580</th>\n",
       "      <td>chrX_155679088_G_C_b38</td>\n",
       "      <td>TTGAACAATGAGATCACTTGGACACAGAGCGGGGAACATCACACAC...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155679088</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>ref</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7581</th>\n",
       "      <td>chrX_155679088_G_C_b38</td>\n",
       "      <td>TTGAACAATGAGATCACTTGGACACAGAGCGGGGAACATCACACAC...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155679088</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>alt</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7582</th>\n",
       "      <td>chrX_155687184_G_T_b38</td>\n",
       "      <td>AAATAGCTGTTCATGCATTCTGTCCACATTTTGAAGTTCAGTTCAG...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155687184</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ref</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7583</th>\n",
       "      <td>chrX_155687184_G_T_b38</td>\n",
       "      <td>AAATAGCTGTTCATGCATTCTGTCCACATTTTGAAGTTCAGTTCAG...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155687184</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>alt</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7584 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  variant_id                                                seq chrom        pos ref alt allele  casual_label\n",
       "0         chr1_63671_G_A_b38  GGGCACCTTCTTTTTCTTAATTGTATCATACATTTTTATTCTGGTC...  chr1      63671   G   A    ref             1\n",
       "1         chr1_63671_G_A_b38  GGGCACCTTCTTTTTCTTAATTGTATCATACATTTTTATTCTGGTC...  chr1      63671   G   A    alt             1\n",
       "2        chr1_108826_G_C_b38  TGGATCCTTAATAAACAAGGATGTTTCTGCATCATTTCCCCACAAC...  chr1     108826   G   C    ref             1\n",
       "3        chr1_108826_G_C_b38  TGGATCCTTAATAAACAAGGATGTTTCTGCATCATTTCCCCACAAC...  chr1     108826   G   C    alt             1\n",
       "4        chr1_135203_G_A_b38  GGAGCTTTGGACTGGGGAGGCCGCAGTGAGGCGAGAGCTAGCTGGG...  chr1     135203   G   A    ref             1\n",
       "...                      ...                                                ...   ...        ...  ..  ..    ...           ...\n",
       "7579  chrX_155668742_C_A_b38  TTTGAGCCTGCTTTAAACCATGAAATTCAACACCATCCTCAGATTC...  chrX  155668742   C   A    alt             0\n",
       "7580  chrX_155679088_G_C_b38  TTGAACAATGAGATCACTTGGACACAGAGCGGGGAACATCACACAC...  chrX  155679088   G   C    ref             0\n",
       "7581  chrX_155679088_G_C_b38  TTGAACAATGAGATCACTTGGACACAGAGCGGGGAACATCACACAC...  chrX  155679088   G   C    alt             0\n",
       "7582  chrX_155687184_G_T_b38  AAATAGCTGTTCATGCATTCTGTCCACATTTTGAAGTTCAGTTCAG...  chrX  155687184   G   T    ref             0\n",
       "7583  chrX_155687184_G_T_b38  AAATAGCTGTTCATGCATTCTGTCCACATTTTGAAGTTCAGTTCAG...  chrX  155687184   G   T    alt             0\n",
       "\n",
       "[7584 rows x 8 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs = []\n",
    "\n",
    "for _, row in tqdm(df.iterrows(), total=len(df)):\n",
    "    chrom = row['chrom']\n",
    "    pos = int(row['pos'])\n",
    "    \n",
    "    start = pos - 100 # pyfaidx 索引从0开始\n",
    "    end = pos + 99\n",
    "    \n",
    "    try:\n",
    "        seq = genome[chrom][start:end].seq.upper()\n",
    "        seqs.append({\n",
    "            'variant_id': variant_id,\n",
    "            'seq': seq,\n",
    "            'chrom': chrom,\n",
    "            'pos': pos,\n",
    "            'ref': ref,\n",
    "            'alt': alt,\n",
    "            'allele': 'ref',\n",
    "            'casual_label': 1,\n",
    "        })\n",
    "        assert seq[99] == ref\n",
    "\n",
    "        alt_seq = seq[:99] + alt + seq[100:]\n",
    "        seqs.append({\n",
    "            'variant_id': variant_id,\n",
    "            'seq': alt_seq,\n",
    "            'chrom': chrom,\n",
    "            'pos': pos,\n",
    "            'ref': ref,\n",
    "            'alt': alt,\n",
    "            'allele': 'alt',\n",
    "            'casual_label': 1,\n",
    "        })\n",
    "\n",
    "\n",
    "    except KeyError:\n",
    "        print(f\"Chromosome {chrom} not found in reference.\")\n",
    "        continue\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "file_name = '../data/GTEx_eQTL/Enformer_processed/data_gtex_fine_vcf_Whole_Blood_neg.vcf'\n",
    "df = pd.read_csv(file_name, sep='\\t', names=['chr', 'pos', 'variant_id', 'ref', 'alt', '1', '2', '3'], comment='#')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "for _, row in tqdm(df.iterrows(), total=len(df)):\n",
    "    # 解析变异ID: chr1_123456_G_A_b37\n",
    "    variant_id = row['variant_id']\n",
    "    chrom, pos, ref, alt, _ = variant_id.split('_')\n",
    "    pos = int(pos)\n",
    "    \n",
    "    # 提取参考序列\n",
    "    start = pos - 100 # pyfaidx 索引从0开始\n",
    "    end = pos + 99\n",
    "    \n",
    "    try:\n",
    "        seq = fasta[chrom][start:end].seq.upper()\n",
    "        seqs.append({\n",
    "            'variant_id': variant_id,\n",
    "            'seq': seq,\n",
    "            'chrom': chrom,\n",
    "            'pos': pos,\n",
    "            'ref': ref,\n",
    "            'alt': alt,\n",
    "            'allele': 'ref',\n",
    "            'casual_label': 0,\n",
    "        })\n",
    "        assert seq[99] == ref\n",
    "\n",
    "        alt_seq = seq[:99] + alt + seq[100:]\n",
    "        seqs.append({\n",
    "            'variant_id': variant_id,\n",
    "            'seq': alt_seq,\n",
    "            'chrom': chrom,\n",
    "            'pos': pos,\n",
    "            'ref': ref,\n",
    "            'alt': alt,\n",
    "            'allele': 'alt',\n",
    "            'casual_label': 0,\n",
    "        })\n",
    "\n",
    "\n",
    "    except KeyError:\n",
    "        print(f\"Chromosome {chrom} not found in reference.\")\n",
    "        continue\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "seq_df = pd.DataFrame(seqs)\n",
    "seq_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_df.to_csv('../data/GTEx_eQTL/Enformer_processed/Whole_Blood_data.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "torch",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
